File: backend/node/api/src/application/use-cases/create-run.ts
import type { Queue } from "../../application/queue/queue";
import { Run } from "../../domain/entities/run";
import { inject } from "../../infra/DI/container";
import type { PipelineRepository } from "../../infra/repository/pipeline";
import type { RunRepository } from "../../infra/repository/run";
import { RunCreatedMapper } from "../mappers/run-created-mapper";

type Input = {
	pipelineId: string;
	payload: number;
};

export class CreateRunUseCase {
	@inject("pipeline-repository")
	private declare readonly pipelineRepository: PipelineRepository;

	@inject("run-repository")
	private declare readonly runRepository: RunRepository;

	@inject("queue")
	private declare readonly queue: Queue;

	async execute(
		input: Input,
	): Promise<[string, undefined] | [undefined, Error]> {
		const [pipeline, pipelineError] = await this.pipelineRepository.getById(
			input.pipelineId,
		);
		if (pipelineError) return [undefined, pipelineError];

		const [run, createError] = Run.create(input);
		if (createError) return [undefined, createError];

		const [, saveError] = await this.runRepository.save(run);
		if (saveError) return [undefined, saveError];

		await this.queue.publish(
			"run.created",
			RunCreatedMapper.toPayload(run, pipeline),
			{ headers: { realtime: true } },
		);

		console.log("passsou????");

		return [run.getId(), undefined];
	}
}

File: backend/go/api/internal/run/use-cases/create-run.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/pipeline"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/queue"
	runPkg "github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/run"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/shared/errors"
)

type CreateRunUseCaseDependencies struct {
	Queue              queue.RabbitMQ
	RunRepository      *runPkg.RunRepository
	PipelineRepository *pipeline.PipelineRepository
}

type CreateRunUseCase struct {
	CreateRunUseCaseDependencies
}

func NewCreateRunUseCase(deps CreateRunUseCaseDependencies) CreateRunUseCase {
	return CreateRunUseCase{
		CreateRunUseCaseDependencies: deps,
	}
}

type CreateRunInput struct {
	PipelineId *string
	Payload    float64
}

func (u *CreateRunUseCase) Execute(input CreateRunInput) error {
	pipeline, err := u.PipelineRepository.GetByID(context.Background(), *input.PipelineId)
	if err != nil {
		return err
	}

	if pipeline == nil {
		return errors.PipelineNotFoundError(*input.PipelineId)
	}

	run, err := runPkg.NewRun(runPkg.NewRunPayload{
		PipelineId: input.PipelineId,
		Payload:    input.Payload,
	})

	if err != nil {
		return err
	}

	err = u.RunRepository.Save(context.Background(), run)
	if err != nil {
		return err
	}

	u.Queue.Publish("run.created", runPkg.RunCreatedMapper(run, pipeline), queue.PublishConfig{})

	return nil
}
